crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -564,6 +564,26 @@ where
             ast::Stmt::Break(_) => {
                 self.loop_break_states.push(self.flow_snapshot());
             }
+
+            ast::Stmt::For(
+                for_stmt @ ast::StmtFor {
+                    range: _,
+                    is_async: _,
+                    target,
+                    iter,
+                    body,
+                    orelse,
+                },
+            ) => {
+                debug_assert!(self.current_assignment.is_none());
+                self.add_standalone_expression(iter);
+                self.visit_expr(iter);
+                self.current_assignment = Some(for_stmt.into());
```

Comment:
Nit: I would move the assertion closer to where we want to ensure that the assignment is null

```suggestion
                self.add_standalone_expression(iter);
                self.visit_expr(iter);
                debug_assert!(self.current_assignment.is_none());
                self.current_assignment = Some(for_stmt.into());
```

Unrelated to this PR: We could do something more fancy if this is an assertion that we want whenever we set an assertion by:

* Adding a `current_assignment(for_stmt.into()) -> AssignmentGuard` method. It asserts that `current_assignment` is `None`
* The `AssignmentGuard` is a wrapper around a [`DropBomb`](https://docs.rs/drop_bomb/latest/drop_bomb/)
* Add a `clear_assignment(guard)` method (or guard.clear(&mut self)`)  that unsets the assignment and defuses the bomb.

Reply:
This existed for a while in the new parser, you can refer to it:

https://github.com/astral-sh/ruff/blob/1475a4d2a8a487a48593a53be33b9f0fac1d1a55/crates/ruff_python_parser/src/parser/mod.rs#L232-L248

Instead of `flags`, it would be `current_assignment: CurrentAssignment`